File: src/schema.ts
/**
 * A-1: herdr socket API schema reflection + parameter validation.
 *
 * Caches `herdr api schema --json` (60s TTL) so herdr_call can validate params
 * against the LIVE schema of the installed herdr binary before sending — no
 * more blind passthrough. `$ref` values are resolved against
 * `schemas.request.$defs` (that is where the daemon publishes param shapes).
 */
import { execSync } from "node:child_process";

export interface ParamsSchema {
  /** Raw property map: name -> JSON schema fragment (may contain $ref/type/enum/anyOf). */
  properties: Record<string, unknown>;
  /** Names required by the schema (top-level `required` on the params def). */
  required: string[];
  /** True when the method's params def is `{type:"object"}` with no properties. */
  empty: boolean;
}

export interface ValidationIssue {
  name: string;
  message: string;
}

export interface ValidationResult {
  ok: boolean;
  errors: ValidationIssue[];
  warnings: ValidationIssue[];
}

const SCHEMA_TTL_MS = 60_000;
let schemaCache: { raw: string; loadedAt: number } | null = null;

interface SchemaDoc {
  schemas?: {
    request?: {
      oneOf?: unknown[];
      $defs?: Record<string, unknown>;
    };
  };
}

function loadSchema(force = false): SchemaDoc {
  const now = Date.now();
  if (!force && schemaCache && now - schemaCache.loadedAt < SCHEMA_TTL_MS) {
    return JSON.parse(schemaCache.raw) as SchemaDoc;
  }
  const raw = execSync("herdr api schema --json", { encoding: "utf-8", timeout: 5000 });
  schemaCache = { raw, loadedAt: now };
  return JSON.parse(raw) as SchemaDoc;
}

function requestDoc(doc: SchemaDoc): { oneOf: unknown[]; defs: Record<string, unknown> } {
  const req = doc?.schemas?.request ?? {};
  return {
    oneOf: Array.isArray(req.oneOf) ? req.oneOf : [],
    defs: (req.$defs ?? {}) as Record<string, unknown>,
  };
}

/** Resolve `#/schemas/request/$defs/Name` (or bare `Name`) against the defs map. */
export function resolveRef(ref: unknown, defs: Record<string, unknown>): Record<string, unknown> | null {
  if (typeof ref !== "string") return null;
  const name = ref.includes("/") ? ref.slice(ref.lastIndexOf("/") + 1) : ref;
  const def = defs[name];
  return def && typeof def === "object" ? (def as Record<string, unknown>) : null;
}

/** Find the params schema for a method by scanning request.oneOf method consts. */
export function getMethodParamsSchema(method: string): ParamsSchema | null {
  const doc = loadSchema();
  const { oneOf, defs } = requestDoc(doc);
  for (const item of oneOf) {
    const rec = (item ?? {}) as Record<string, unknown>;
    const props = (rec["properties"] ?? {}) as Record<string, unknown>;
    const mprop = (props["method"] ?? {}) as Record<string, unknown>;
    if (mprop["const"] !== method) continue;
    const pprops = (props["params"] ?? {}) as Record<string, unknown>;
    // Resolve the params $ref (most methods) or inline shape.
    const paramsDef = resolveRef(pprops["$ref"], defs) ?? pprops;
    const propMap = (paramsDef["properties"] ?? {}) as Record<string, unknown>;
    const required = Array.isArray(paramsDef["required"]) ? (paramsDef["required"] as string[]) : [];
    const empty = Object.keys(propMap).length === 0 && (paramsDef["type"] === "object" || paramsDef["type"] === undefined);
    return { properties: propMap, required, empty };
  }
  return null;
}
/** List every method name with its params schema INLINED (resolve $ref to
 * properties + required) so clients see param names without guessing. */
export function listMethods(query = ""): { method: string; params: { properties: Record<string, unknown>; required: string[]; empty: boolean } }[] {
  const doc = loadSchema();
  const { oneOf } = requestDoc(doc);
  const out: { method: string; params: { properties: Record<string, unknown>; required: string[]; empty: boolean } }[] = [];
  for (const item of oneOf) {
    const rec = (item ?? {}) as Record<string, unknown>;
    const props = (rec["properties"] ?? {}) as Record<string, unknown>;
    const mprop = (props["method"] ?? {}) as Record<string, unknown>;
    const method = typeof mprop["const"] === "string" ? mprop["const"] : null;
    if (!method || (query && !method.toLowerCase().includes(query.toLowerCase()))) continue;
    out.push({ method, params: getMethodParamsSchema(method) ?? { properties: {}, required: [], empty: true } });
  }
  return out;
}

function primitiveType(prop: unknown, defs: Record<string, unknown>): { type: string | string[] | null; enum?: unknown[] } | null {
  const rec = (prop ?? {}) as Record<string, unknown>;
  if (typeof rec["type"] === "string" || Array.isArray(rec["type"])) {
    return { type: rec["type"] as string | string[], enum: Array.isArray(rec["enum"]) ? (rec["enum"] as unknown[]) : undefined };
  }
  if (Array.isArray(rec["enum"])) return { type: "string", enum: rec["enum"] as unknown[] };
  const refd = rec["$ref"] !== undefined ? resolveRef(rec["$ref"], defs) : null;
  if (refd) {
    if (Array.isArray(refd["enum"])) return { type: "string", enum: refd["enum"] as unknown[] };
    if (typeof refd["type"] === "string" || Array.isArray(refd["type"])) return { type: refd["type"] as string | string[], enum: Array.isArray(refd["enum"]) ? (refd["enum"] as unknown[]) : undefined };
  }
  if (Array.isArray(rec["anyOf"])) return null; // composite — skip type check
  return null;
}

function matchesType(value: unknown, types: string | string[]): boolean {
  const list = Array.isArray(types) ? types : [types];
  for (const t of list) {
    switch (t) {
      case "string": if (typeof value === "string") return true; break;
      case "number": case "integer": if (typeof value === "number") return true; break;
      case "boolean": if (typeof value === "boolean") return true; break;
      case "object": if (value !== null && typeof value === "object") return true; break;
      case "array": if (Array.isArray(value)) return true; break;
      case "null": if (value === null) return true; break;
    }
  }
  return false;
}

/**
 * Validate `params` for `method` against the live schema. Unknown params are
 * WARNINGS (schema may lag the daemon); missing required / wrong type / wrong
 * enum are ERRORS. Returns ok:false without touching the socket when errors.
 */
export function validateMethodParams(method: string, params: Record<string, unknown>): ValidationResult {
  const errors: ValidationIssue[] = [];
  const warnings: ValidationIssue[] = [];
  const ps = getMethodParamsSchema(method);
  if (!ps) {
    // Method unknown to schema — cannot validate; treat as pass-through warning.
    warnings.push({ name: "method", message: `method "${method}" not found in herdr schema — passed through unvalidated` });
    return { ok: true, errors, warnings };
  }
  if (ps.empty) {
    const extra = Object.keys(params ?? {});
    if (extra.length > 0) warnings.push({ name: "params", message: `method takes no params; got: ${extra.join(", ")}` });
    return { ok: true, errors, warnings };
  }
  const doc = loadSchema();
  const { defs } = requestDoc(doc);
  const given = params ?? {};

  for (const name of ps.required) {
    if (!(name in given) || given[name] === undefined) {
      errors.push({ name, message: `missing required param "${name}"` });
    }
  }
  for (const [name, value] of Object.entries(given)) {
    const prop = ps.properties[name];
    if (!prop) {
      warnings.push({ name, message: `unknown param "${name}" (not in schema — daemon may still accept it)` });
      continue;
    }
    const pt = primitiveType(prop, defs);
    if (!pt || value === undefined) continue;
    if (pt.type !== null && value !== null && !matchesType(value, pt.type)) {
      errors.push({ name, message: `param "${name}" should be ${Array.isArray(pt.type) ? pt.type.join("|") : pt.type}, got ${typeof value}` });
      continue;
    }
    if (Array.isArray(pt.enum) && typeof value === "string" && !pt.enum.includes(value)) {
      errors.push({ name, message: `param "${name}" must be one of [${pt.enum.join(", ")}], got "${value}"` });
    }
  }
  return { ok: errors.length === 0, errors, warnings };
}

File: crates/herdr-mcp/src/schema.rs
use serde_json::{Map, Value};
use std::io::Read;
use std::process::{Command, Stdio};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex, OnceLock};
use std::thread;
use std::time::{Duration, Instant};

const SCHEMA_TTL: Duration = Duration::from_secs(60);
const SCHEMA_LOAD_TIMEOUT: Duration = Duration::from_secs(8);
const MAX_SCHEMA_BYTES: usize = 8 * 1024 * 1024;

static CACHE: OnceLock<Mutex<Option<CachedSchema>>> = OnceLock::new();
static REFRESHING: AtomicBool = AtomicBool::new(false);

#[derive(Debug, Clone)]
struct CachedSchema {
    loaded_at: Instant,
    registry: Arc<SchemaRegistry>,
}

#[derive(Debug, Clone)]
pub struct SchemaRegistry {
    methods: Vec<MethodSchema>,
    defs: Map<String, Value>,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct MethodSchema {
    pub method: String,
    pub properties: Map<String, Value>,
    pub required: Vec<String>,
    pub empty: bool,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct ValidationIssue {
    pub name: String,
    pub message: String,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct ValidationResult {
    pub ok: bool,
    pub errors: Vec<ValidationIssue>,
    pub warnings: Vec<ValidationIssue>,
}

pub fn list_methods(query: &str) -> Result<Vec<MethodSchema>, String> {
    let registry = load_registry(false)?;
    let query = query.to_ascii_lowercase();
    Ok(registry
        .methods
        .iter()
        .filter(|method| query.is_empty() || method.method.to_ascii_lowercase().contains(&query))
        .cloned()
        .collect())
}

pub fn validate_method_params(method: &str, params: &Value) -> Result<ValidationResult, String> {
    let registry = load_registry(false)?;
    Ok(validate_with_registry(&registry, method, params))
}

/// Warm the live Herdr schema without delaying server startup.
pub fn prewarm_async() {
    let cache = CACHE.get_or_init(|| Mutex::new(None));
    if cache
        .lock()
        .ok()
        .is_some_and(|guard| guard.as_ref().is_some())
    {
        return;
    }
    start_background_refresh();
}

fn validate_with_registry(
    registry: &SchemaRegistry,
    method: &str,
    params: &Value,
) -> ValidationResult {
    let mut errors = Vec::new();
    let mut warnings = Vec::new();
    let Some(schema) = registry
        .methods
        .iter()
        .find(|candidate| candidate.method == method)
    else {
        warnings.push(ValidationIssue {
            name: "method".to_owned(),
            message: format!(
                "method \"{method}\" not found in herdr schema - passed through unvalidated"
            ),
        });
        return ValidationResult {
            ok: true,
            errors,
            warnings,
        };
    };

    let Some(given) = params.as_object() else {
        errors.push(ValidationIssue {
            name: "params".to_owned(),
            message: "params must be a JSON object".to_owned(),
        });
        return ValidationResult {
            ok: false,
            errors,
            warnings,
        };
    };

    if schema.empty {
        if !given.is_empty() {
            warnings.push(ValidationIssue {
                name: "params".to_owned(),
                message: format!(
                    "method takes no params; got: {}",
                    given.keys().cloned().collect::<Vec<_>>().join(", ")
                ),
            });
        }
        return ValidationResult {
            ok: true,
            errors,
            warnings,
        };
    }

    for name in &schema.required {
        if !given.contains_key(name) {
            errors.push(ValidationIssue {
                name: name.clone(),
                message: format!("missing required param \"{name}\""),
            });
        }
    }

    for (name, value) in given {
        let Some(property) = schema.properties.get(name) else {
            errors.push(ValidationIssue {
                name: name.clone(),
                message: format!("unknown param \"{name}\" (not in schema)"),
            });
            continue;
        };
        let Some(primitive) = primitive_type(property, &registry.defs) else {
            continue;
        };
        if !value.is_null() && !matches_type(value, &primitive.types) {
            errors.push(ValidationIssue {
                name: name.clone(),
                message: format!(
                    "param \"{name}\" should be {}, got {}",
                    primitive.types.join("|"),
                    value_type(value)
                ),
            });
            continue;
        }
        if let Some(allowed) = primitive.allowed
            && value.is_string()
            && !allowed.iter().any(|candidate| candidate == value)
        {
            errors.push(ValidationIssue {
                name: name.clone(),
                message: format!(
                    "param \"{name}\" must be one of [{}], got {}",
                    allowed
                        .iter()
                        .map(Value::to_string)
                        .collect::<Vec<_>>()
                        .join(", "),
                    value
                ),
            });
        }
    }

    ValidationResult {
        ok: errors.is_empty(),
        errors,
        warnings,
    }
}

#[derive(Debug)]
struct PrimitiveType {
    types: Vec<String>,
    allowed: Option<Vec<Value>>,
}

fn primitive_type(property: &Value, defs: &Map<String, Value>) -> Option<PrimitiveType> {
    let object = property.as_object()?;
    if object.get("anyOf").is_some_and(Value::is_array) {
        return None;
    }

    let resolved = object
        .get("$ref")
        .and_then(Value::as_str)
        .and_then(|reference| resolve_ref(reference, defs))
        .and_then(Value::as_object);
    let effective = resolved.unwrap_or(object);

    let allowed = effective.get("enum").and_then(Value::as_array).cloned();
    let types = match effective.get("type") {
        Some(Value::String(value)) => vec![value.clone()],
        Some(Value::Array(values)) => values
            .iter()
            .filter_map(Value::as_str)
            .map(str::to_owned)
            .collect(),
        _ if allowed.is_some() => vec!["string".to_owned()],
        _ => return None,
    };

    Some(PrimitiveType { types, allowed })
}

fn matches_type(value: &Value, types: &[String]) -> bool {
    types.iter().any(|kind| match kind.as_str() {
        "string" => value.is_string(),
        "number" => value.is_number(),
        "integer" => value
            .as_i64()
            .map(|_| true)
            .or_else(|| value.as_u64().map(|_| true))
            .unwrap_or(false),
        "boolean" => value.is_boolean(),
        "object" => value.is_object(),
        "array" => value.is_array(),
        "null" => value.is_null(),
        _ => true,
    })
}

fn value_type(value: &Value) -> &'static str {
    match value {
        Value::Null => "null",
        Value::Bool(_) => "boolean",
        Value::Number(number) if number.is_i64() || number.is_u64() => "integer",
        Value::Number(_) => "number",
        Value::String(_) => "string",
        Value::Array(_) => "array",
        Value::Object(_) => "object",
    }
}

fn load_registry(force: bool) -> Result<Arc<SchemaRegistry>, String> {
    let cache = CACHE.get_or_init(|| Mutex::new(None));
    {
        let guard = cache
            .lock()
            .map_err(|_| "schema cache lock poisoned".to_owned())?;
        if !force && let Some(cached) = guard.as_ref() {
            let registry = Arc::clone(&cached.registry);
            if cached.loaded_at.elapsed() < SCHEMA_TTL {
                return Ok(registry);
            }
            drop(guard);
            start_background_refresh();
            return Ok(registry);
        }
    }

    match load_live_registry() {
        Ok(registry) => {
            let registry = Arc::new(registry);
            let mut guard = cache
                .lock()
                .map_err(|_| "schema cache lock poisoned".to_owned())?;
            *guard = Some(CachedSchema {
                loaded_at: Instant::now(),
                registry: Arc::clone(&registry),
            });
            Ok(registry)
        }
        Err(error) => {
            let guard = cache
                .lock()
                .map_err(|_| "schema cache lock poisoned".to_owned())?;
            if let Some(cached) = guard.as_ref() {
                Ok(Arc::clone(&cached.registry))
            } else {
                Err(error)
            }
        }
    }
}

fn start_background_refresh() {
    if REFRESHING
        .compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire)
        .is_err()
    {
        return;
    }
    thread::spawn(|| {
        if let Ok(registry) = load_live_registry()
            && let Ok(mut guard) = CACHE.get_or_init(|| Mutex::new(None)).lock()
        {
            *guard = Some(CachedSchema {
                loaded_at: Instant::now(),
                registry: Arc::new(registry),
            });
        }
        REFRESHING.store(false, Ordering::Release);
    });
}

fn load_live_registry() -> Result<SchemaRegistry, String> {
    let raw = run_schema_command()?;
    let document: Value = serde_json::from_slice(&raw)
        .map_err(|error| format!("cannot parse `herdr api schema --json`: {error}"))?;
    registry_from_document(&document)
}

fn run_schema_command() -> Result<Vec<u8>, String> {
    let mut child = Command::new("herdr")
        .args(["api", "schema", "--json"])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .map_err(|error| format!("cannot start `herdr api schema --json`: {error}"))?;

    let stdout = child
        .stdout
        .take()
        .ok_or_else(|| "cannot capture herdr schema stdout".to_owned())?;
    let stderr = child
        .stderr
        .take()
        .ok_or_else(|| "cannot capture herdr schema stderr".to_owned())?;
    let stdout_reader = thread::spawn(move || read_stream(stdout));
    let stderr_reader = thread::spawn(move || read_stream(stderr));
    let started = Instant::now();

    let status = loop {
        match child
            .try_wait()
            .map_err(|error| format!("cannot wait for herdr schema command: {error}"))?
        {
            Some(status) => break status,
            None if started.elapsed() >= SCHEMA_LOAD_TIMEOUT => {
                let _ = child.kill();
                let _ = child.wait();
                let _ = stdout_reader.join();
                let _ = stderr_reader.join();
                return Err(format!(
                    "`herdr api schema --json` exceeded {}ms",
                    SCHEMA_LOAD_TIMEOUT.as_millis()
                ));
            }
            None => thread::sleep(Duration::from_millis(20)),
        }
    };

    let stdout = stdout_reader
        .join()
        .map_err(|_| "herdr schema stdout reader panicked".to_owned())??;
    let stderr = stderr_reader
        .join()
        .map_err(|_| "herdr schema stderr reader panicked".to_owned())??;
    if !status.success() {
        return Err(format!(
            "`herdr api schema --json` failed: {}",
            String::from_utf8_lossy(&stderr).trim()
        ));
    }
    if stdout.len() > MAX_SCHEMA_BYTES {
        return Err(format!("herdr schema exceeded {MAX_SCHEMA_BYTES} bytes"));
    }
    Ok(stdout)
}

fn read_stream(mut stream: impl Read) -> Result<Vec<u8>, String> {
    let mut output = Vec::new();
    stream
        .read_to_end(&mut output)
        .map_err(|error| format!("cannot read herdr schema process output: {error}"))?;
    Ok(output)
}

fn registry_from_document(document: &Value) -> Result<SchemaRegistry, String> {
    let request = document
        .pointer("/schemas/request")
        .and_then(Value::as_object)
        .ok_or_else(|| "herdr schema missing schemas.request".to_owned())?;
    let one_of = request
        .get("oneOf")
        .and_then(Value::as_array)
        .ok_or_else(|| "herdr schema missing schemas.request.oneOf".to_owned())?;
    let defs = request
        .get("$defs")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();
    let mut methods = Vec::new();

    for item in one_of {
        let Some(properties) = item.get("properties").and_then(Value::as_object) else {
            continue;
        };
        let Some(method) = properties
            .get("method")
            .and_then(|value| value.get("const"))
            .and_then(Value::as_str)
        else {
            continue;
        };
        let params_shape = properties
            .get("params")
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default();
        let resolved = params_shape
            .get("$ref")
            .and_then(Value::as_str)
            .and_then(|reference| resolve_ref(reference, &defs))
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or(params_shape);
        let param_properties = resolved
            .get("properties")
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default();
        let required = resolved
            .get("required")
            .and_then(Value::as_array)
            .map(|values| {
                values
                    .iter()
                    .filter_map(Value::as_str)
                    .map(str::to_owned)
                    .collect::<Vec<_>>()
            })
            .unwrap_or_default();
        let empty = param_properties.is_empty()
            && matches!(
                resolved.get("type").and_then(Value::as_str),
                None | Some("object")
            );

        methods.push(MethodSchema {
            method: method.to_owned(),
            properties: param_properties,
            required,
            empty,
        });
    }

    if methods.is_empty() {
        return Err("herdr schema contains no request methods".to_owned());
    }
    Ok(SchemaRegistry { methods, defs })
}

fn resolve_ref<'a>(reference: &str, defs: &'a Map<String, Value>) -> Option<&'a Value> {
    let name = reference.rsplit('/').next().unwrap_or(reference);
    defs.get(name)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn registry() -> SchemaRegistry {
        let document = json!({
            "schemas": {
                "request": {
                    "oneOf": [
                        {
                            "properties": {
                                "method": {"const": "ping"},
                                "params": {"$ref": "#/schemas/request/$defs/PingParams"}
                            }
                        },
                        {
                            "properties": {
                                "method": {"const": "agent.start"},
                                "params": {"$ref": "#/schemas/request/$defs/AgentStartParams"}
                            }
                        }
                    ],
                    "$defs": {
                        "PingParams": {"type": "object", "properties": {}},
                        "AgentStartParams": {
                            "type": "object",
                            "properties": {
                                "workspace_id": {"type": "string"},
                                "count": {"type": "integer"},
                                "kind": {"enum": ["pi", "grok"]}
                            },
                            "required": ["workspace_id"]
                        }
                    }
                }
            }
        });
        registry_from_document(&document).unwrap()
    }

    #[test]
    fn parses_method_param_schema() {
        let registry = registry();
        assert_eq!(registry.methods.len(), 2);
        assert!(registry.methods[0].empty);
        assert_eq!(registry.methods[1].required, vec!["workspace_id"]);
        assert!(registry.methods[1].properties.contains_key("kind"));
    }

    #[test]
    fn validates_required_type_enum_and_unknown_params() {
        let registry = registry();
        let valid = validate_with_registry(
            &registry,
            "agent.start",
            &json!({"workspace_id": "w1", "count": 2, "kind": "pi", "future": true}),
        );
        assert!(!valid.ok);
        assert_eq!(valid.errors.len(), 1);
        assert!(valid.warnings.is_empty());

        let invalid = validate_with_registry(
            &registry,
            "agent.start",
            &json!({"count": "two", "kind": "other"}),
        );
        assert!(!invalid.ok);
        assert_eq!(invalid.errors.len(), 3);
    }

    #[test]
    fn unknown_method_passes_with_warning() {
        let result = validate_with_registry(&registry(), "future.method", &json!({}));
        assert!(result.ok);
        assert_eq!(result.warnings.len(), 1);
    }
}
